### ISSUE TITLE
Add progress reporting where appropriate

### ISSUE THREAD

**Issue by Arlodotexe** [Arlodotexe/OwlCore.Storage#54]
Progress reporting is a good call out. Generally the library is built so that progress reporting can be implemented application-side during an operation (e.g. navigating directories, moving things around, stream transfers). However, there's a few long-running extension methods where we could consider an `IProgress` parameter, such as Copy and Move. 

[Here](https://github.com/Arlodotexe/OwlCore.Storage/blob/c0aa68cb8fdf8856369403d03eb07f0edd83fbc7/src/Extensions/CopyAndMoveExtensions.cs#L44) is the inbox implementation for copying a file into a folder. Presumably, dotnet should supply us with an overload that takes an `IProgress` in `Stream.CopyToAsync`, but that doesn't appear to be the case. We'd need to build this ourselves or use one of the many versions that others have built.

_Originally posted by @Arlodotexe in https://github.com/Arlodotexe/OwlCore.Storage/issues/12#issuecomment-2160903901_

**Comment by itsWindows11** [comment]
ChatGPT proposes this:

```cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

public async Task CopyStreamWithProgressAsync(
    Stream source, 
    Stream destination, 
    IProgress<double> progress = null, 
    CancellationToken cancellationToken = default)
{
    const int bufferSize = 81920; // 80 KB buffer
    byte[] buffer = new byte[bufferSize];
    long totalBytesRead = 0;
    long totalLength = source.Length;

    int bytesRead;
    while ((bytesRead = await source.ReadAsync(buffer, 0, buffer.Length, cancellationToken)) > 0)
    {
        await destination.WriteAsync(buffer, 0, bytesRead, cancellationToken);
        totalBytesRead += bytesRead;
        
        if (progress != null && totalLength > 0)
        {
            double percentage = (double)totalBytesRead / totalLength * 100;
            progress.Report(percentage);
        }
    }
}
```

Though we need to report the total bytes read along with the length instead of the percentage, and leave the rest for the implementor to decide what they want to do with the progress, and use `ArrayPool`. No idea if some underlying stream implementations will require special handling here.

**Comment by Arlodotexe** [comment]
Yeah, we should look at what others have already brainstormed online on sites like StackOverflow.

**Comment by Arlodotexe** [comment]
We also need to expose the `bufferSize` parameter in `CreateCopyOfAsync` (interface, and extension and fallback delegate). Since they're similar breaking changes, they should ship together.

**Comment by Arlodotexe** [comment]
Added note.

Looking at the trajectory and what we picked up coming from #108 regarding aligning properties with other storables, handling property watching, and responsibly contracting disposability, the stabilized shape of the abstraction hinted that "copy/move progress reporting" and "file watching" overlap.

Projecting ahead, if we bring forward the generalizations that were enabled by aligning file watching with progress reporting, look at the overlap that progress reporting brings (total moved vs max movable) in contrast to file watching, then carry those implications back into *folder* watching, then we may find that the resulting API surface has a shape that naturally lends itself to a "storage transfer manager" with minimal changes.
